### ISSUE TITLE
Support for CUDA aware MPI

### ISSUE THREAD

**Issue by vchuravy** [JuliaParallel/MPI.jl#219]
Just capturing a discussion with @lcw I had a while ago.

A more and more common usecase is to combine CUDA and MPI, and some MPI implementations allow for RDMA to CUDA pointers.

The pattern one usually sees is
```
comm_buffer = ...
# recv data
copyto!(A, comm_buffer)
 ```

Where `A` is a `CuArray` and `comm_buffer` could either be a host array allocated in pinned memory or with CUDA-aware MPI could be a `CuArray` itself.

1. Add support for pinned memory e.g. `cudaMallocHost` to speed up the pattern.
2. Detect CUDA-aware MPI configuration and allow the usage of CUDA pointers directly.

This feature doesn't really fit MPI.jl or Cu*.jl, so we might want to create a glue repository that takes care of this.

- https://devblogs.nvidia.com/benchmarking-cuda-aware-mpi/
- https://github.com/parallel-forall/code-samples/tree/master/posts/cuda-aware-mpi-example/src

**Comment by lcw** [comment]
@vchuravy, thanks for summarizing this.

I am not sure where the CUDA aware MPI should sit.

But the support for pinned memory could possibly sit in CUDAdrv.jl.  It would be useful for out-of-core gnu calculations (for example) in addition to using with CUDA unaware MPI.

**Comment by vchuravy** [comment]
Basic implementation

```julia
import MPI
using CUDAdrv
using CuArrays
function MPI.Isend(buf::CuArray{T}, dest::Integer, tag::Integer,
                            comm::MPI.Comm) where T
    GC.@preserve buf begin
        _buf = Base.cconvert(CuPtr{T}, buf)
        cuptr = Base.unsafe_convert(CuPtr{T}, _buf)
        ptr = reinterpret(Ptr{T}, cuptr)
        MPI.Isend(ptr, length(buf), dest, tag, comm)
    end
end

function MPI.Irecv!(buf::CuArray{T}, src::Integer, tag::Integer,
    comm::MPI.Comm) where T
    GC.@preserve buf begin
        _buf = Base.cconvert(CuPtr{T}, buf)
        cuptr = Base.unsafe_convert(CuPtr{T}, _buf)
        ptr = reinterpret(Ptr{T}, cuptr)
        MPI.Irecv!(ptr, length(buf), src, tag, comm)
    end
end

```

**Comment by simonbyrne** [comment]
Closed by #302
